src/main/haskell/kore/src/Kore/Step/Simplification/AndTerms.hs

```diff
@@ -0,0 +1,499 @@
+
+Returns NotHandled if it could not handle the input.
+-}
+equalInjectiveHeadsAnd
+    ::  ( MetaOrObject level
+        , Hashable variable
+        , IntVariable variable
+        , Ord (variable level)
+        , Ord (variable Meta)
+        , Ord (variable Object)
+        , Show (variable level)
+        , SortedVariable variable
+        )
+    => MetadataTools level StepperAttributes
+    -> TermSimplifier level variable
+    -- ^ Used to simplify subterm "and".
+    -> PureMLPattern level variable
+    -> PureMLPattern level variable
+    -> FunctionResult
+        (Maybe
+            ( IntCounter (ExpandedPattern level variable
+            , SimplificationProof level)
+            )
+        )
+equalInjectiveHeadsAnd
+    tools
+    termMerger
+    (App_ firstHead firstChildren)
+    (App_ secondHead secondChildren)
+  | StepperAttributes.isInjective firstHeadAttributes
+    && StepperAttributes.isInjective secondHeadAttributes
+    && firstHead == secondHead
+  = Handled $ do -- Maybe monad
+    intCounterChildren <- sequenceA $
+        zipWith termMerger firstChildren secondChildren
+    return $ do -- IntCounter monad
+        children <- sequenceA intCounterChildren
+        (   PredicateSubstitution
+                { predicate = mergedPredicate
+                , substitution = mergedSubstitution
+                }
+            , _proof
+            ) <- mergePredicatesAndSubstitutions
+                tools
+                (map (ExpandedPattern.predicate . fst) children)
+                (map (ExpandedPattern.substitution . fst) children)
+        return
+            ( ExpandedPattern
+                { term = give (MetadataTools.sortTools tools) $
+                    mkApp firstHead (map (ExpandedPattern.term . fst) children)
+                , predicate = mergedPredicate
+                , substitution = mergedSubstitution
+                }
+            , SimplificationProof
+            )
+  where
+    firstHeadAttributes = MetadataTools.symAttributes tools firstHead
+    secondHeadAttributes = MetadataTools.symAttributes tools firstHead
+equalInjectiveHeadsAnd _ _ _ _ = NotHandled
+
+{-| And simplification for patterns with sortInjection heads.
+
+Assumes that the two heads were already tested for equality and were found
+to be different.
+
+Returns NotHandled if it could not handle the input.
+-}
+sortInjectionAndAssumesDifferentHeads
+    ::  ( Eq (variable Object)
+        , MetaOrObject level
+        )
+    => MetadataTools level StepperAttributes
+    -> PureMLPattern level variable
+    -> PureMLPattern level variable
+    -> FunctionResult (PureMLPattern level variable, SimplificationProof level)
+sortInjectionAndAssumesDifferentHeads
+    tools
+    (App_ firstHead _)
+    (App_ secondHead _)
+  | StepperAttributes.isSortInjection firstHeadAttributes
+    && StepperAttributes.isSortInjection secondHeadAttributes
+  =
+    assert (firstHead /= secondHead) $
+        Handled (mkBottom, SimplificationProof)
```

Comment:
This is not always so.  to be sure of bottom you must be sure that none of the of the injected terms could be topped in a sort injection  (maybe after some further simplifications).